File: highest_product_of_3.rb
ary_of_ints = [1, -5, -10, 6, -11, 8]
# highest_product_of_3(ary_of_ints)

def highest_product_of_3(ary_of_ints)
  highest_product_of_3 = ary_of_ints[0] * ary_of_ints[1] * ary_of_ints[2]

  highest_product_of_2 = ary_of_ints[0] * ary_of_ints[1]
  lowest_product_of_2 = ary_of_ints[0] * ary_of_ints[1]

  highest_int = ary_of_ints.first(2).max
  lowest_int = ary_of_ints.first(2).min

  ary_of_ints.each_with_index do |current, index|
    next if index < 2

    highest_product_of_3 = [
      highest_product_of_3,
      highest_product_of_2 * current,
      lowest_product_of_2 * current
    ].max

    highest_product_of_2 = [
      highest_product_of_2,
      highest_int * current,
      lowest_int * current
    ].max

    lowest_product_of_2 = [
      lowest_product_of_2,
      highest_int * current,
      lowest_int * current
    ].min

    highest_int = [highest_int, current].max

    lowest_int = [lowest_int, current].min
  end

  highest_product_of_3
end

puts highest_product_of_3(ary_of_ints)

File: highestProductOf3.js
var arrayOfInts = [1, -5, -10, 6, -11, 8];
// highestProductOf3(arrayOfInts)

function highestProductOf3(arrayOfInts) {
  var highestProductOf3 = arrayOfInts[0] * arrayOfInts[1] * arrayOfInts[2],
      highestProductOf2 = arrayOfInts[0] * arrayOfInts[1],
      lowestProductOf2  = arrayOfInts[0] * arrayOfInts[1],
      highest = Math.max(arrayOfInts[0], arrayOfInts[1]),
      lowest = Math.min(arrayOfInts[0], arrayOfInts[1]);


  for (i = 2; i < arrayOfInts.length; i++) {
    var current = arrayOfInts[i];

    highestProductOf3 = Math.max(
      highestProductOf3,
      highestProductOf2 * current,
      lowestProductOf2 * current
    );


    highestProductOf2
    lowestProductOf2
    highest
    lowest
  }

};
